Update page Develop -> Testing -> Benchmarking (#305)
polkadot-developers/polkadot-docs: develop/parachains/testing/benchmarking.md

Reviewer: Minor fix for SEO proposes the description has 162 instead of the max 160, besides that I made it more SEO friendly 
```suggestion
Learn how to use FRAME's benchmarking framework to evaluate custom pallets, ensure accurate weights, and optimize on-chain computations and extrinsics.

### Prepare Your Environment

> Install the [`frame-omni-bencher`](https://crates.io/crates/frame-omni-bencher){target=\_blank} command-line tool:
> ```bash

Reviewer 2: Just to ensure consistency with the rest of the page:

```suggestion
Install the [`frame-omni-bencher`](https://crates.io/crates/frame-omni-bencher){target=\_blank} command-line tool:

```bash
```

> - **Use the `#[extrinsic_call]` macro** - this macro is used when calling the extrinsic itself and is a required part of a benchmarking function. See the [`extrinsic_call Rust docs](https://paritytech.github.io/polkadot-sdk/master/frame_benchmarking/v2/index.html#extrinsic_call-and-block){target=_blank} for more details

Reviewer 2: ```suggestion
- **Use the `#[extrinsic_call]` macro** - this macro is used when calling the extrinsic itself and is a required part of a benchmarking function. See the [`extrinsic_call`](https://paritytech.github.io/polkadot-sdk/master/frame_benchmarking/v2/index.html#extrinsic_call-and-block){target=\_blank} docs for more details
```

- **Validate extrinsic behavior** - the `assert_eq` expression ensures that the extrinsic is working properly within the benchmark context

> Add the `benchmarking` module to your pallet. In the pallet `lib.rs` file add the following:
> ```rust
> #[cfg(feature = "runtime-benchmarks")]
> mod benchmarking;
> ```

Reviewer 2: Same here

```suggestion
Add the `benchmarking` module to your pallet. In the pallet `lib.rs` file add the following:

```rust
#[cfg(feature = "runtime-benchmarks")]
mod benchmarking;
```
```

### Add Benchmarks to Runtime

Before running the benchmarking tool, you must integrate benchmarks with your runtime as follows:

1. Navigate to your `runtime/src` directory and check if a `benchmarks.rs` file exists. If not, create one. This file will contain the macro that registers all pallets for benchmarking along with their respective configurations:
    ```rust title="benchmarks.rs"
    --8<-- 'code/develop/parachains/testing/benchmarking/frame-benchmark-macro.rs'
    ```

    For example, to add a new pallet named `pallet_parachain_template` for benchmarking, include it in the macro as shown:
    ```rust title="benchmarks.rs" hl_lines="3"
    --8<-- 'code/develop/parachains/testing/benchmarking/frame-benchmark-macro.rs::3'
    );
    ```

    !!!warning "Updating `define_benchmarks!` macro is required"

>         Any pallet that needs to be benchmarked must be included in the `define_benchmarks!` macro. The CLI will only be able to access and benchmark pallets that are registered here.

Reviewer 2: ```suggestion
        Any pallet that needs to be benchmarked must be included in the [`define_benchmarks!`](https://paritytech.github.io/polkadot-sdk/master/frame_benchmarking/macro.define_benchmarks.html){target=\_blank} macro. The CLI will only be able to access and benchmark pallets that are registered here.
```

Reviewer 3: Very nice (the rewording and adding the link)

    The `runtime-benchmarks` feature gate ensures benchmark tests are isolated from production runtime code.

3. Enable runtime benchmarking for your pallet in `runtime/Cargo.toml`:
    ```toml
    --8<-- 'code/develop/parachains/testing/benchmarking/runtime-cargo.toml'
    ```

### Run Benchmarks

You can now compile your runtime with the `runtime-benchmarks` feature flag. This feature flag is crucial as the benchmarking tool will look for this feature being enabled to know when it should run benchmark tests. Follow these steps to compile the runtime with benchmarking enabled:

1. Run `build` with the feature flag included:

    ```bash
    cargo build --features runtime-benchmarks --release
    ```

> 2. Create a `weights.rs` file in your pallet's `src/` directory. This file will store the auto-generated weight calculations:

Author: Yeah, it is weird. It crashed if I didn't create the file before

--output ./pallets/benchmarking/frame-weight-template.hbs
```

4. Run the benchmarking tool to measure extrinsic weights:

    ```bash
    frame-omni-bencher v1 benchmark pallet \
    --runtime INSERT_PATH_TO_WASM_RUNTIME \
    --pallet INSERT_NAME_OF_PALLET \
    --extrinsic "" \
    --template ./frame-weight-template.hbs \
    --output weights.rs
    ```

    !!! info "Flag definitions"
        - `--runtime` - the path to your runtime's Wasm
        - `--pallet` - the name of the pallet you wish to benchmark. This pallet must be configured in your runtime and defined in `define_benchmarks`
        - `--extrinsic` - which extrinsic to test. Using `""` implies all extrinsics will be benchmarked
        - `--template` - defines how weight information should be formatted
        - `--output` - where the output of the auto-generated weights will reside

The generated `weights.rs` file contains weight annotations for your extrinsics, ready to be added to your pallet. The output should be similar to the following. Some output is omitted for brevity:

--8<-- 'code/develop/parachains/testing/benchmarking/benchmark-output.html'

#### Add Benchmark Weights to Pallet

> Once the `weights.rs` is generated, you need to integrate it with your pallet. This involves two steps: importing the `weights` module and the `WeightInfo` trait, and adding it to your pallet's `Config` trait. Here's how to set this up:

Reviewer 3: ```suggestion
Once the `weights.rs` is generated, you must integrate it with your pallet. To begin the integration, import the `weights` module and the `WeightInfo` trait, then add both to your pallet's `Config` trait. Complete the following steps to set up the configuration:
```
Please add a bullet here for "1. (describe this first step) above the first rust code snippet

> ```rust

Reviewer 3: ```suggestion
```rust title="INSERT_FILE_NAME_HERE"
```

--8<-- 'code/develop/parachains/testing/benchmarking/weight-config.rs'
```

> After which, you must add this to the `#[pallet::weight]` annotation in all the extrinsics via the `Config`:

Reviewer 3: ```suggestion
2. Next, you must add this to the `#[pallet::weight]` annotation in all the extrinsics via the `Config` as follows:
```

> ```rust hl_lines="2"

Reviewer 3: ```suggestion
```rust hl_lines="2" title="INSERT_FILE_NAME_HERE"
```

--8<-- 'code/develop/parachains/testing/benchmarking/dispatchable-pallet-weight.rs'
```

> Finally, configure the actual weight values in your runtime. In `runtime/src/config/mod.rs`, add:

Reviewer 3: ```suggestion
3. Finally, configure the actual weight values in your runtime. In `runtime/src/config/mod.rs`, add the following code:
```

> ```rust

Reviewer 3: ```suggestion
```rust title="mod.rs"
```
